### DISCUSSION TITLE
setup.py install_requires vs. requirements.txt

### DISCUSSION THREAD

**Discussion by frafra** [ckan/ckan#6601] [Ideas]
It is not possible to install most of CKAN extensions because they are not packaged as regular Python libraries. That means that pip sees no dependencies, and a different requirements file (its name could vary from repository to repository) has to be downloaded and passed to pip. These requirements files contain pinned dependencies, so the last extension which has been installed win over the others, in terms of dependencies. This totally hides potential incompatibilities, makes builds dependent on the order in which extensions have been installed, and make it more complex to install and handle extensions than it should.

Following standard Python practices for libraries makes handling extensions a terrible experience. Couldn't we just follow Python standards and avoid such dependencies' roulette?

Explained here too: https://github.com/ckan/ckan/discussions/6601#discussioncomment-1814292

https://github.com/ckan/ckanext-spatial/blob/9831e74c9e9eedcb35fa00858190f778d83928ad/setup.py#L46

**Comment by Zharktas** [comment]
None of the core extension have, you are supposed to install them from requirements.txt as there might be conficts between extensions which pip can't resolve

**Reply by categulario** [reply]
as @frafra points out, it is the system administrator's responsibility to pin the dependencies that work for _their_ setup and environment, not the project's. If two packages have conflicting dependencies then they are incompatible and the user should know that as soon as possible, not when things randomly break later

**Comment by frafra** [comment]
This is getting quite messy in my opinion, as each extension now has its own specific way to be installed. For example, I need to look for `requirements.txt` sometimes, `requirements-py2.txt` some other times, or just `setup.py` for some other extensions.

Shouldn't `setup.py` contain dependencies like `lib1>=1.0.0` instead of pinned dependencies like `lib1==1.2.3` while `requirements.txt` and similar are used to pin the dependencies version? If something fails to install, then it really means there is an incompatibility. There is no way of figuring out if a set of extension is compatible with one another, and the order used to install the extension produces a different result.

Wouldn't be better to just follow Python standards as every other published Python package does?

**Comment by Zharktas** [comment]
Can't seem to find the previous discussion about it, but in general when you have multiple extensions from various sources, using install_requires the install will fail if there is a conflict between versions. Using requirements.txt, you can always solve the issue manually after successful install. It is even in the [recommendations](https://docs.ckan.org/en/2.9/extensions/best-practices.html#add-third-party-libraries-to-requirements-txt). 

But this comes up from time to time and there is no actual solution except following ckan conventions.

**Reply by frafra** [reply]
This is not correct, as pinned dependencies should not stay in install_requires. We have pinned dependencies in many requirements.txt, which bring conflicts. There is a difference between relaxed constraints used in libraries (declaring "I am compatible with Flask 2.x") and pinned constraints used in softwares to improve reproducibility ("CKAN has been tested with Flask 2.1.0"). `install_requires` should declare the requirements with a wider range of compatibility. requirements.txt can still be used for pinned requirements if needed. We are just ignoring compatibilities at the moment, and the last extension which gets installed, overrides the dependencies of the previously installed extension. It is messy, and widely different from what happens from Python libraries on PyPI, and it is still a non-standard and terrible practice even if it is suggested by CKAN documentation, in my opinion.

**Comment by frafra** [comment]
Thanks, @Zharktas, for taking the time to reply.

It is not `install_requires` which makes the installation fail, but conflicting requirements, which is what should happen. It is useful to know if something has been declared as incompatible by the developer. This is not something that should be hidden. I think that CKAN guidelines are based on a misunderstanding on how packaging is commonly done in Python packages.

CKAN core extensions fail to handle incompatibilities, breaks compatibility with regular Python tooling, and clashes against Python packaging guidelines and conventions.

Here some examples of problems caused by CKAN policy:
- It is not possible for a developer to explicitly declare that an extension is compatible only with some version of a dependency
- Incompatibilities are ignored and hidden
- Unexpected issues after CKAN is installed or deployed because incompatibilities were never specified
- Incompatibilities can arise (again, without user being aware of them) depending on the order used to install the extension (installing A and then B produces a different environment compared to installing B and then A)
- Cannot install extensions with pip (dependencies are not specified in `setup.py`)
- Cannot distribute extensions (dependencies are not specified in `setup.py`)
- Incompatibility with Python tools and ecosystems (no Poetry, PDM, or any other tool to handle dependencies, conflicts and such)
- Breaking compatibility because requirements files may be renamed (it happened with the `*-py2.txt` files)
- Custom procedures and code to handle CKAN extensions

I see no reason why CKAN should not follow Python guidelines and the existing standard to declare dependencies. `setup.py` should contain a list of dependencies (requires B > 1.2), while requirements is for pinning dependencies for complete projects and/or for CI (install B == 1.3.2).

> Whereas install_requires requirements are minimal, requirements files often contain an exhaustive listing of pinned versions for the purpose of achieving repeatable installations of a complete environment.

> It is not considered best practice to use install_requires to pin dependencies to specific versions, or to specify sub-dependencies (i.e. dependencies of your dependencies). This is overly-restrictive, and prevents the user from gaining the benefit of dependency upgrades.

-- https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/#install-requires

I would then suggest aligning with what is the current standard in the Python community. That would make things way more reliable and robust, which is something CKAN could greatly benefit from.

**Comment by Zharktas** [comment]
I moved this to ckan discussions as it has more to do what is the recommendations in ckan extensions than a single extension itself.

**Reply by frafra** [reply]
Thanks, I understand that policy changes should be discussed here, especially since there are multiple repositories affected. I hope the proposal will not die here.
I think that a Python module that is supposed to be installed with pip that doesn't declare its dependencies is a bug itself, though.

**Comment by EricSoroos** [comment]
I had a discussion (somewhere, the searches aren't finding one authored by me) where I've pointed out something similar, how basically right now the last extension installed wins the dependency lottery, and how our configuration is preventing the tools we have (like pipdeptree) from being able to help us even detect the conflicts.

**Reply by frafra** [reply]
Every kind of Python dependencies tool rely on what is described in `install_requires`, to figure out how to satisfy dependencies in a stable and reliable way. We need to have relaxed dependencies (instead of pinned ones) specified in `install_requires`, as every other Python library does.

**Comment by EricSoroos** [comment]
I've found an initial version on my drive, It must be in a closed discussion somewhere that's not indexed. 

**CKAN version**

All currently shipping

**Describe the bug**

The current CKAN coding standards for extensions leads to inconsistent installs and works against the zen of the python packaging system. 

Specifically: (https://docs.ckan.org/en/2.9/extensions/best-practices.html#add-third-party-libraries-to-requirements-txt)
> If your extension requires third party libraries, rather than adding them to setup.py, they should be added to requirements.txt, which can be installed with: `pip install -r requirements.txt`

Separating the dependencies from setup.py breaks the pip resolver. So not only are we likely to get conflicts, we can't even use the tools to figure out what they are. 

**Steps to reproduce**

A small example:
* Install ckan. A pinned version of redis is installed. 
* Install ckanext-harvest. The original version of redis is uninstalled and a different version of redis is installed. 

I'm sure that there are other examples, e.g requests, six.

**Expected behavior**

Ideally, the ckan, extensions, and all of the python dependencies could be installed with: `pip install --use-feature=2020-resolver ckan ckanext_foo ckanext_bar ckanext_scheming ckanext_pages...`

This would be a one-shot install, and the pip resolver could build the entire dependency tree, and inform us of conflicts, which could then be resolved.

Then, we could see the results of our dependencies with `pipdeptree`:

```
$ pipdeptree
Warning!! Cyclic dependencies found:
* SPARQLWrapper => rdflib => SPARQLWrapper
* rdflib => SPARQLWrapper => rdflib
------------------------------------------------------------------------
bleach==2.1.3
  - html5lib [required: >=0.99999999pre,!=1.0b8,!=1.0b7,!=1.0b6,!=1.0b5,!=1.0b4,!=1.0b3,!=1.0b2,!=1.0b1, installed: 1.0.1]
    - six [required: >=1.9, installed: 1.15.0]
    - webencodings [required: Any, installed: 0.5.1]
  - six [required: Any, installed: 1.15.0]
ckan==2.8.7
ckanext-dcat==1.1.1
ckanext-harvest==1.3.2
  - ckantoolkit [required: ==0.0.3, installed: 0.0.3]
  - pika [required: >=1.1.0, installed: 1.2.0]
  - redis [required: <=3.0, installed: 2.10.6]
  - requests [required: >=2.11.1, installed: 2.25.1]
    - certifi [required: >=2017.4.17, installed: 2020.12.5]
    - chardet [required: >=3.0.2,<5, installed: 4.0.0]
    - idna [required: >=2.5,<3, installed: 2.10]
    - urllib3 [required: >=1.21.1,<1.27, installed: 1.26.3]
  - six [required: >=1.12.0, installed: 1.15.0]
...
```

Where now, CKAN has no dependencies. (note, ckanext-harvest has had it's requirements hoisted (and tweaked) into setup.py in my install.)

**Additional details**


* To reduce the number of conflicts, extensions should be liberal in the versions what dependencies they pin, e.g. pinning a specific version of requests is unnecessary, especially given how stable the requests interface is. 

```
requests>=2,<3
```

Obviously, there are cases where a specific pinned dependency is required. That's to be expected if you are relying on the specific behavior of a dependency Transitive dependencies should not be pinned. (e.g., certifi and pytz have stable interfaces, and should generally _always_ be on the latest version). Any extension dependency that's also a dependency of CKAN shouldn't be pinned, because then the CKAN pins for different major versions will conflict. 

CKAN should be the most specific pin, and extensions compatible with a specific version of CKAN should be compatible with that pin. However, CKAN should also be generally forgiving of the exact version of dependencies that have relatively stable interfaces where it's not using 'deep' features. 

Ideally, CKAN would also be only pinning direct dependencies to make it easier to have a clear view of what needs to be updated on point releases. But that's a further bridge to cross.

**Comment by categulario** [comment]
Just to contribute my 2 cents on this discussion.

I think the benefits are very well explained by @EricSoroos and @frafra and the opposing argument is just "these are the ckan guidelines" but no technical argument (that hasn't been addressed) remains. Guidelines can and should be changed if the project as a whole can benefit from it.

**Reply by frafra** [reply]
Thanks for your message. I tried to further explain what the problems are in the first message and reply to another message, to see if we can fix this.

**Comment by smotornyuk** [comment]
+1, all my extensions use loose dependencies(just like Ian did in ckanext-scheming) and I like it. But it likely will cause issues like #6415. No matter how reliable and stable CKAN dependencies are, there always will be a chance, that a new release contains a bug. 
With the current release process, it may take quite a bit of time till the next CKAN patch release(~ half a year), which will mark the broken dependency version as inappropriate. So we need more frequent releases in order to handle such issues as soon as possible. This brings us to #6381 :)

**Reply by frafra** [reply]
I totally agree with you.
CKAN itself should use loose dependencies, in my opinion, and generate pinned ones on a regular basis, to be used for releases, builds, etc.

I do not see using loose dependencies as a regression for the issue you mentioned, because in both scenarios the user has to do an extra step to intervene on a specific dependency (does not really change if that is a requirement file o a temporary one liner).

**Reply by wardi** [reply]
I forgot that @TkTech made the change to `install_requires` for ckanext-scheming. It does seem to work for this extension because there are very few dependencies.

**Comment by wardi** [comment]
Fully agree with the desire to have easily installed extensions and "normal" python packaging. Here's some background on why this hasn't worked for us in the past:

## `install_requires` is enforced `requirements.txt` is a suggestion
Where we use `install_requires` we need to specify a version range of every dependency or pip will prevent the installation of anything deemed to conflict by another package. This is a real problem for integrators even if they "know" some combination will work (or simply want to test) the tool prevents them from doing so.

Some ckan-related projects like ckanapi do use `install_requires` with version ranges because it helps end users and the number of dependencies are limited.

Versions pinned in `requirements.txt` only apply if this is the first time a dependency is being `pip install`ed. This is the dependency "roulette" mentioned. It's fragile because it depends on the order an integrator installs the `requirements.txt` files.

`requirements.txt` is the current recommended approach for extensions because
## semantic versioning is aspirational
version ranges are really hard to get right, and dependency upgrades break working code even with bugfix or minor point releases. For `install_requires` both of these problems fall to the maintainers of the ckan extensions. These maintainers have made the effort to publish and update their code but are typically only paid to actually maintain their sites with their specific versions of everything, so supporting users that run into issues on different versions of ckan or with different combinations of extensions falls to the wayside.

It's much easier for volunteer-maintained shared extensions to publish a list of dependencies that are known to work with their code (a `requirements.txt`) and given that most volunteer extension authors are rarely even keeping up with the current version of ckan or fixing serious bugs, asking them to debug every combination of dependency version issue is unlikely to succeed.

## integrators are responsible for their environments
CKAN system integrators *are* paid and responsible for the sites they deliver to clients. They are best positioned to resolve dependencies for their particular combination of ckan version and extensions.

This could involve writing a script that runs `pip install requirements.txt` in the "right" order for their site or by maintaining a separate pinned set of requirements. This is finicky and unrewarding work, but asking the extension maintainers to do this for every possible combination of extensions is unlikely to succeed.

## a better future
These are hard problems that stem from competing interests and limited available time from volunteer extension maintainers and system integrators. The inflexibility of `install_requires` even when an integrator "knows better" or when a developer simply wants to test an upgrade is another problem with recommending this route.

If we can cut through this somehow to make ckan and its extensions easier to install, that would be great for our users. Maybe with a small number of well-supported extensions we could start to use `install_requires` if we special-cased or limited the number of supported ckan versions. I'm happy to see experiments in this direction to see how well it can work.

**Reply by wardi** [reply]
For ckan 2.10 + we're Python 3 only. Maybe that makes dependencies less of a minefield than they have been when we were trying to hang on to versions of dependencies (and dependencies of dependencies) that still worked with Python 2

**Reply by EricSoroos** [reply]
One significant difference with the past is that pip now has a much better dependency resolver (since `--use-feature=2020-resolver` in Oct 2020, and current py3 only pip doesn't even have the old option.). Previously, even if there were two compatible ranges specified, pip would error out with a dependency resolution error, where now it will generally find a solution. 

FWIW, Many of the extensions I'm using are actually lightweight forks, adding the install_requires and a minimal MANIFEST.in* to enable a simple `pip install ckanext-foo` to work as a one would expect from a modern python package.  For example, my ckanext-spatial (with somewhat updated pins) is:
```
    install_requires=[
            'GeoAlchemy2>=0.8.0',
            'Shapely>=1.7',
            'pyproj==2.2.2; python_version < "3.0"',
            'pyproj==2.6.1; python_version > "3.0"',
            'OWSLib==0.18.0; python_version < "3.0"',
            'OWSLib>=0.21.0; python_version > "3.0"',
            'lxml>=2.3',
            'argparse',
            'pyparsing>=2.1.10',
            'requests<=3',
    ],
```


(*) This is usually enough of a MANIFEST.in.
```
graft ckanext
include *.py
```

**Reply by frafra** [reply]
> Fully agree with the desire to have easily installed extensions and "normal" python packaging. Here's some background on why this hasn't worked for us in the past:

Thank you for giving us some context :)

> Where we use `install_requires` we need to specify a version range of every dependency or pip will prevent the installation of anything deemed to conflict by another package. This is a real problem for integrators even if they "know" some combination will work (or simply want to test) the tool prevents them from doing so.

If a module also work with a version of a dependency which is not included in `install_requires`, that is a bug. Better having to fix some false positives than have no clue about compatibilities between versions, right?

> version ranges are really hard to get right [...]

They are not if you have tests. If you do not have tests because you consider your extension to be very simple, then you are not going to have many dependencies, probably.
 
> It's much easier for volunteer-maintained shared extensions to publish a list of dependencies that are known to work with their code [...]
> CKAN system integrators _are_ paid and responsible for the sites they deliver to clients. They are best positioned to resolve dependencies for their particular combination of ckan version and extensions.

How should a CKAN system integrator figure out which dependencies should be used? Why should such a person be in a better position to know which dependencies should be used instead of a developer? He/she would need to run tests, or just install and hope for the best, until some obscure bug appears.
Couldn't we just run these test on the CI instead of doing the same work multiple times?

I would greatly prefer to open pull requests to improve dependencies ranges listed in `install_requires` than have to figure out everything by myself and having no change to share it.

Would you imagine if Django apps (just to make an example) would use `install_requires` and `requirements.txt` like CKAN extension do today? Or if Python libraries in general would follow the same approach? I think I would have nightmares each time I would need to add a dependency :)

> If we can cut through this somehow to make ckan and its extensions easier to install, that would be great for our users. Maybe with a small number of well-supported extensions we could start to use `install_requires` if we special-cased or limited the number of supported ckan versions. I'm happy to see experiments in this direction to see how well it can work.

I am in favour of a gradual approach to reach a better solution, so we can discuss, apply and test better solutions for CKAN, and extend them when we consider they prove to work well :)

**Comment by amercader** [comment]
To move this discussion forward, why don't we set up a repo that runs one or more workflows on Github Actions where we try to install at least some (or all) of the extensions in the CKAN organization (see #6787 for a list) using one of the methods mentioned (eg loose `install_requires`) and see if we actually find issues?

Would that help?

**Reply by frafra** [reply]
What I do is to extract all the dependencies in a common format, make a list and see if they are compatible.

```Earthfile
requirements-auto:
    DO +INSTALL --pkgs="bsdmainutils"
    COPY --dir scripts .
    FOR ext IN dcat kata oaipmh oauth2 spatial
        COPY ckanext/ckanext-${ext}/requirements.txt ckanext-${ext}.txt
    END
    FOR ext IN harvest
        COPY ckanext/ckanext-${ext}/pip-requirements.txt ckanext-${ext}.txt
    END
    FOR ext IN doi scheming
        COPY ckanext/ckanext-${ext}/setup.py ckanext-${ext}.py
    END
    RUN scripts/gather-requirements.sh ckanext-* > requirements.in
    RUN sed -i -r -f scripts/fix_requirements.sed requirements.in
    SAVE ARTIFACT requirements.in

requirements:
    DO +INSTALL_PY --pkgs="pip-tools"
    COPY +requirements-auto/requirements.in .
    RUN pip-compile --no-annotate --no-header \
            --output-file=requirements.txt \
            requirements.in \
            $CKAN_VENV/src/ckan/requirements.txt
    SAVE ARTIFACT requirements.txt
```

This is how I extract requirements from a `setup.py` file:
```python
#!/usr/bin/env python3

import ast
import sys

path = sys.argv[1]
parsed = ast.parse(open(path).read())
for node in parsed.body:
    if not isinstance(node, ast.Expr):
        continue
    if not isinstance(node.value, ast.Call):
        continue
    if node.value.func.id != "setup":
        continue
    for keyword in node.value.keywords:
        if keyword.arg == "install_requires":
            requirements = ast.literal_eval(keyword.value)
            print("\n".join(requirements))
```

While this is the sed expression file which I fiddle with to manually address incompatibilities. Here is an example:

```sed
# relax requirement for ckanext-dcat
s/^(rdflib==.*) (#.*)$/#\1\2/
```

We could have a GitHub action trying to run pip-compile on a list of dependencies which is generated with similar procedures, so we can spot incompatibilities and address them properly. Then we could try to run all the extensions together, but if we have no conflict, they should just work. Loosening requirements would be more an improvement for future releases and a change of the guidelines and practices.

**Comment by pdelboca** [comment]
I agree on moving forward to a pilot experience. This would be a nice ticket to focus as a part of `CKAN 3.0` since, imho, I wouldn't delay `2.10` for this change.

If I Understood Correctly, next steps to try this could be:
 - Move `requirements.txt` into `setup.py`
 - Select some core extensions from #6787 and migrate them to the new requirement system.
 - Create a new repository (or a reproducible example) of how the new install process would work.
   - We can do a first iteration with just 2 or 3: scheming, xloader, harvest.
   - Check-in
   - Complete the list on #6787
 - Review the repository/example and, if agree on moving forward, update CKAN documentation to reflect this new practice. This will require to:
   - Update CKAN's documentation to reflect this new practice
   - Update extension's documentation on how to handle requirements
   - Perhaps updating our cookiecutter template to reflect this somehow.

One of the links in the [Python documentation](https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/#install-requires-vs-requirements-files) provides some useful insight (at least it was clarifying for me) on how can we structure our projects: https://caremad.io/posts/2013/07/setup-vs-requirement/

Concretelly, the example of the `requirements.txt` file:
```
--index-url https://pypi.python.org/simple/

-e https://github.com/foo/bar.git#egg=bar
-e .
```

Can be use to provide some kind of "backwards compatibility" for all the current implementations that relies on `pip install -r requirements.txt` for the setup.

@frafra @EricSoroos what do you think about my proposed "roadmap" for this? Since you are heavily invested into this and have already make some work I think it's logic for you to develop an example/repository. Wanna take the lead?

**Reply by frafra** [reply]
I like the idea of keeping backwards compatibility.

How would such an example repository look like?

**Comment by pdelboca** [comment]
While doing a migration of a big instance to CKAN 2.10 (Yes, I know I'm ahead but it is providing useful insight) I came across with some requirements issues described in this Thread.

For my surprise I found a comment made by myself that I didn't remember 😛 

We are having end of year breaks and with the upcoming release of CKAN 2.10 we're full of work  but I'm gonna revamp this conversation and implementation as one of the first things to do after CKAN 2.10 release. I consider that it will not only improve the quality of CKAN but also could provide a good boundary and definition on what we support.

My proposed goal for the following release (2.10.1 or whatever) will be to complete the requirements update listed in https://github.com/ckan/ckan/issues/7083 and set a new standard to set our requirements. (But this is off topic and again, getting ahead)

@frafra To answer your old question: I have been working on two monitor repositories: https://github.com/ckan/ckan-extensions-monitor and https://github.com/ckan/ckan-python-monitor and I think this could be a good approach.

A simple workflow pulling and installing 2 or 3 extensions from branches with the new way of defining the requirements will be enough to begin with.
